src/tapedfunction.jl

```diff
@@ -168,127 +154,180 @@ function (instr::Instruction{F})() where F
         instr.output.val = output
         instr.tape.counter += 1
     catch e
+        println("counter=", instr.tape.counter)
+        println("tape=", instr.tape)
         println(e, catch_backtrace());
         rethrow(e);
     end
 end
 
+function (instr::GotoInstruction)()
+    if !val(instr.condition) # unless
+        instr.tape.counter = instr.dest
+    else
+        instr.tape.counter += 1
+    end
+end
+
+function (instr::ReturnInstruction)()
+    instr.tape.retval = val(instr.arg)
+end
+
+
+## internal functions
+
+_accurate_typeof(v) = typeof(v)
+_accurate_typeof(v::Type) = Type{v}
+
 """
-    __new__(T, args)
+    __new__(T, args...)
 
 Return a new instance of `T` with `args` even when there is no inner constructor for these args.
 Source: https://discourse.julialang.org/t/create-a-struct-with-uninitialized-fields/6967/5
 """
-@generated function __new__(T, args)
+@generated function __new__(T, args...)
     return Expr(:splatnew, :T, :args)
 end
 
-function (instr::Instruction{typeof(__new__)})()
-    # catch run-time exceptions / errors.
-    try
-        input = map(val, instr.input)
-        T = input[1]
-        args = input[2:end]
-        output = __new__(T, args)
-
-        instr.output.val = output
-        instr.tape.counter += 1
-    catch e
-        println(e, catch_backtrace());
-        rethrow(e);
-    end
-end
 
-function (instr::BlockInstruction)()
-    instr.tape.counter += 1
-end
+## Translation: CodeInfo -> Tape
 
-function (instr::BranchInstruction)()
-    if instr.condition === nothing || !val(instr.condition) # unless
-        target = instr.block_id
-        target_idx = instr.tape.block_map[target]
-        blk_ins = instr.tape[target_idx]
-        @assert isa(blk_ins, BlockInstruction)
-        @assert length(instr.args) == length(blk_ins.args)
-        for i in 1:length(instr.args)
-            blk_ins.args[i].val = val(instr.args[i])
-        end
-        instr.tape.counter = target_idx
-    else
-        instr.tape.counter += 1
-    end
-end
+arg_boxer(var, boxes::Dict{Symbol, Box{<:Any}}) = var
+arg_boxer(var::Core.SSAValue, boxes::Dict{Symbol, Box{<:Any}}) = arg_boxer(Symbol(var.id), boxes)
+arg_boxer(var::Core.TypedSlot, boxes::Dict{Symbol, Box{<:Any}}) =
+    arg_boxer(Symbol("_$(var.id)"), boxes)
+arg_boxer(var::Core.SlotNumber, boxes::Dict{Symbol, Box{<:Any}}) =
+    arg_boxer(Symbol("_$(var.id)"), boxes)
+arg_boxer(var::Symbol, boxes::Dict{Symbol, Box{<:Any}}) =
+    get!(boxes, var, Box{Any}(var, nothing))
```

Comment:
Not sure how much it helps but one can avoid constructing the default value in every call:

```suggestion
    get!(boxes, var) do
        return Box{Any}(var, nothing)
    end
```